Mathlib/Data/Matrix/Rank.lean

```diff
@@ -274,4 +275,39 @@ theorem rank_eq_finrank_span_row [LinearOrderedField R] [Finite m] (A : Matrix m
   rw [← rank_transpose, rank_eq_finrank_span_cols, transpose_transpose]
 #align matrix.rank_eq_finrank_span_row Matrix.rank_eq_finrank_span_row
 
+section MulNonsingInv
+
+/-! ### Lemmas about pre or post multiplying by an invertible matrix. -/
+
+/-- Pre (Left) multiplying by an invertible matrix does not change the rank -/
+lemma rank_mul_IsUnit {m n R: Type}
+  [Fintype n][DecidableEq n][CommRing R]
+  (A: Matrix n n R)(B: Matrix m n R)(hA: IsUnit A.det):
+  (B⬝A).rank = B.rank := by
+  rw [Matrix.rank, mulVecLin_mul, LinearMap.range_comp_of_range_eq_top, ←Matrix.rank ]
+  rw [LinearMap.range_eq_top]
+  intros x
+  use ((A⁻¹).mulVecLin x)
+  rw [mulVecLin_apply, mulVecLin_apply, mulVec_mulVec, mul_nonsing_inv, one_mulVec]
+  exact hA
+
+/-- Post (right) multiplying by an invertible matrix does not change the rank -/
+lemma rank_IsUnit_mul {m n R: Type}
+  [Fintype m][Fintype n][DecidableEq m][Field R]
+  (A: Matrix m m R)(B: Matrix m n R)(hA: IsUnit A.det):
+  (A⬝B).rank = B.rank := by
+  suffices h: LinearMap.ker ((A⬝B).mulVecLin) = LinearMap.ker (B.mulVecLin)
+  have hB := LinearMap.finrank_range_add_finrank_ker (B.mulVecLin)
+  rw [← LinearMap.finrank_range_add_finrank_ker ((A⬝B).mulVecLin), h, add_left_inj] at hB
+  rw [Matrix.rank, Matrix.rank, hB]
+  rw [mulVecLin_mul, LinearMap.ker_comp_of_ker_eq_bot]
+  rw [LinearMap.ker_eq_bot]
+  simp_rw [Function.Injective, mulVecLin_apply]
+  intros x y h
+  replace h := congr_arg (fun x => Matrix.mulVec (A⁻¹) x)  h
```

Comment:
I think the tactic `apply_fun` might be useful here, as `apply_fun Matrix.mulVec (A⁻¹) at h` (also I'm not sure if the brackets are needed.)